feat(api): Add basic API with mock message formats and no crypto.

rfk opened the pull request:
This PR takes the learnings from [1] and tries to turn them into a minimal first step for this repo.

The code here does not implement any crypto at all, and it does not implement any of the TLS1.3 wire format.  But what it does do is *look the same* from the caller's point of view:

* It works with a binary wire format, like a TLS-based implementation would (but much simpler!)
* It uses the same state-machine as a TLS-based implementation would.
* It sends the same sequence of handshake messages as a TLS-based implementation would (albiet with no crypto!)

I'm posting it to get initial feedback on how we feel about this thing from an API consumer level.  Would an API shaped like this "fit" into our existing work without too much hassle?  If not, can we tweak it while maintaining the handshake flow and state-machine?

If it does seem promising, I think we should go ahead and publish this minimal mock version of the API, so that we can work on integrating it into the existing pairing code in parallel with digging further into the crypto bits.

To try it out, you can open `./demo/test_client.html` and `./demo/test_server.html` in two parallel browser windows.  These use a combination of the websocket channelserver and this mock TLS layer to construct a mock-encrypted chat between two windows, reminiscent of our pairing flow.

@eoger @vladikoff @shane-tomlinson f?

Connects to #3.

eoger commented:
From an API user point of view, this is what I imagined so I'm pretty happy with what I see.
Once `initialize` resolves, is the SSL handshake completed?

rfk commented:
> Once initialize resolves, is the TLS handshake completed?

No, although it will resolve transparently as you try to use the `Connection` object.  The handshake can't resolve until the app makes at least one call to `recv` in order to feed to handshake data into the state-machine.  (This is why `recv()` can sometimes return `null`, if the received message contained only handshake data and no app-level data).

shane-tomlinson commented:
> IIUC, Fenix is expected to target quite old generations of Android. 

Following on, I always forget that Fenix is going to use Gecko under the hood so async/await won't be a problem there, we'll control both sides (desktop and mobile). However, if we hope to target apps like Lockbox or browsers like Firefox on FireTV that use the OS level Webview, then we do have to be mindful of which syntax we use.

eoger commented:
@shane-tomlinson would we load that async/await babel polyfill in the whole content bundle or would it be only loaded for the pairing scenarios? @vladikoff says we'll probably do the same as the crypto reliers (separate from the "main" bundle).
In the latter case I think it doesn't matter too much and we should not block on this.

rfk commented on lib/index.js:
```diff
@@ -0,0 +1,532 @@
+    }
+
+    async initialize(psk, psk_id) {
+        this.psk = coerceToBytes(psk) // XXX TODO: probably the app should pass this in as a Uint8Array.
```
We should make it so, good call.

eoger replied:
Would it be useful to leverage the type system here so we don't call initialize twice? e.g. make `initialize` static and return an instance or make it return another class instance?
It might be too much of a refactor so I'm also happy with `if initialized then throw boom` :)

rfk replied:
I am a big fan of using the type system for things, let's give it a shot...

The only reason these aren't arguments to the constructor is because this function wants to be `async`, so a static async initializer probably makes sense.

rfk commented:
> I don't understand the complexities involved in any depth, but naively, ISTM that if we stick to
> "simple" uses of await then we should be able to mechanically convert them into promise chains
> without requiring any additional runtime support

Some projects claim to be able to do this:

* https://www.npmjs.com/package/babel-plugin-transform-async-to-promises
* https://www.npmjs.com/package/babel-plugin-async-to-promises

rfk commented:
@vladikoff I've integrated your build-and-test work, done some refactoring to split things up into more manageable files, and added some basic tests.  I think this is ready for an r? as a preliminary version for testing out the API.

shane-tomlinson commented:
> I definitely understand the concerns here, but I also submit that clarity of code is extra important for thorny crypto code like this. 

Agreed, and thanks for looking up those transformation modules. Would it make sense to define a "minimum UA requirements", a feature-level SLA we can point to? If our SLA explicitly states we are only going to support UAs with async/await support, then no transforms will be needed.

vladikoff commented on demo/tlschannel.js:
```diff
@@ -0,0 +1,101 @@
+class InsecureTLSChannel {
+
+    constructor() {
+        this.socket = null
```
no semicolons in this? I think we should >_>

rfk replied:
The auth-server has ruined my ability to be consistent with semicolons; we should probably add in eslint here to keep us on track.

vladikoff commented on src/crypto.js:
```diff
@@ -0,0 +1,14 @@
+// In this file we will implement the "key schedule" from
+// https://tools.ietf.org/html/rfc8446#section-7.1, which
+// defines how to calculate various keys as the handshake
+// state progreses.
```
typo: `progresses`

vladikoff commented on src/statemachine/index.js:
```diff
@@ -0,0 +1,119 @@
+}
+
+// A special "error" state for when something goes wrong.
+// This state never transitions to another state, effecitvely
```
`effecitvely`

vladikoff commented on src/statemachine/index.js:
```diff
@@ -0,0 +1,119 @@
+    }
+}
+
+// The "connected" state, for when the handshake is compelte
```
typo: `compelte`

vladikoff commented on src/index.js:
```diff
@@ -0,0 +1,182 @@
+//    // Send data to the server by calling `send`,
+//    // which will use the callback provided in the constructor.
+//    // A single `send()` by the application may result in multiple
+//    // invokations of the callback.
```
`invocations` ?

vladikoff commented on src/index.js:
```diff
@@ -0,0 +1,182 @@
+//    await conn.close()
+//
+// The `ServerConnection` API operates similarly; the distinction is mainly
+// in which side is expected to send vs receieve during the protocol handshake.
```
typo: `receieve`

vladikoff commented on src/index.js:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    // These are the three public API methods that
+    // consumers can use to connunicate over TLS.
```
typo: `connunicate`

vladikoff commented on src/messages.js:
```diff
@@ -0,0 +1,80 @@
+// each message type.  We're hoping it's manageable, but need to work
+// through a lot of details.
+//
+// For now, we just send the raw data, prefixied with a length field
```
typo: `prefixed`

vladikoff commented on demo/tlschannel.js:
```diff
@@ -0,0 +1,101 @@
+            const onChannelEcho = async event => {
+                try {
+                  // The channelserver will echo back the provided channelId,
+                  // which lets us know we're successully connected.
```
typo: `successully`

vladikoff commented on demo/tlschannel.js:
```diff
@@ -0,0 +1,101 @@
+        this.socket = new WebSocket(uri);
+        // Once the initial channelserver setup is done, push all data
+        // received from websocket into the TLS state-machine to decrypt it,
+        // then feed it back out via our own `onReceieve` callback.
```
typo `onReceieve`

vladikoff commented on src/index.js:
```diff
@@ -0,0 +1,182 @@
+// `ServerConnection` classes.  They each take a callback to be used for
+// sending data to the remote peer, and operate like this:
+//
+//    conn = await ClientConnection.create(psk, pskID, async function send_data_to_server(data) {
```
we can maybe use example syntax like https://github.com/mozilla/fxa-crypto-relier/blob/master/src/deriver/ScopedKeys.js#L19 , maybe

vladikoff replied:
`pskID` -> `pskId` ?

vladikoff commented on src/index.js:
```diff
@@ -0,0 +1,182 @@
+
+    // Subclasses will override this with some async initialization logic.
+    static async create(psk, pskID, sendCallback) {
+        return new this(psk, pskID, sendCallback)
```
do you want to assert `sendCallback` ? as a must have?

vladikoff commented on demo/tlschannel.js:
```diff
@@ -0,0 +1,101 @@
+        return new Promise((resolve, reject) => {
+            const onChannelCreation = async event => {
+                try {
+                    // Receive the channelID for the newly-created channel.
```
`channelId`

vladikoff commented on demo/tlschannel.js:
```diff
@@ -0,0 +1,101 @@
+                try {
+                    // Receive the channelID for the newly-created channel.
+                    self.socket.removeEventListener('message', onChannelCreation)
+                    const { channelid } = JSON.parse(event.data)
```
channelId ?(casing)

vladikoff replied:
do we want to throw on json parse fail?

rfk replied:
yes we do; I *think* this should throw automatically and get routed to `reject`...

vladikoff commented on src/index.js:
```diff
@@ -0,0 +1,182 @@
+    // XXX TODO: this risks keeping the result of the last call around
+    // in memory, we should probably "void" it out somehow on completion.
+
+    _serialize(cb) {
```
is `_serialize` the right name for this? 🤔  I wonder if I am missing something

rfk replied:
I mean it in the sense of "one thing after another in strict order" but am open to better names.

rfk replied:
@vladikoff perhaps `_synchronized` would be a better name, by analogy to Java's `synchronized` keyword?

vladikoff replied:
forgot to respond here, yes `synchronized` ++

rfk commented on webpack.config.js:
```diff
@@ -24,13 +30,11 @@ if (env === 'min') {
+      path: __dirname + '/dist',
       filename: `tls.${min}js`,
-      library: 'fxaTls',
+      library: 'tls',
```
That makes sense, I'll revert it.

rfk commented on demo/tlschannel.js:
```diff
@@ -0,0 +1,101 @@
+                    const { channelid } = JSON.parse(event.data)
+                    // The socket's ready, we can now initialize the TLS state-machine.
+                    self.channelId = channelid
+                    self.tlsconn = await tls.InsecureServerConnection.create(FAKE_PSK, FAKE_PSK_ID, data => {
```
yep

rfk replied:
The new version requires them to be generated or passed in rather than using fixed values.

rfk commented on demo/tlschannel.js:
```diff
@@ -0,0 +1,101 @@
+        const self = new this()
+        self._makeSocket(CHANNEL_SERVER)
+        return new Promise((resolve, reject) => {
+            const onChannelCreation = async event => {
```
thanks, I've imported the eslint config from fxa-content-server and will do a big cleanup on formatting and conventions here.

rfk commented on demo/tlschannel.js:
```diff
@@ -0,0 +1,101 @@
+                    self.channelId = channelId
+                    self.tlsconn = await tls.InsecureClientConnection.create(FAKE_PSK, FAKE_PSK_ID, data => {
+                        // To send data over the websocket, it needs to be encoded as a safe string.
+                        self.socket.send(tls.bytesToHex(data))
```
I think it may actually be channelserver server that requires we send b64urlsafe data over this particular channel.

rfk commented:
Thanks for the feedback all!  I've changed enough things here in addressing feedback and expanding the capabilities that I think it's better to put up a fresh PR, let's continue over in https://github.com/mozilla/fxa-pairing-tls/pull/8
